components/pattern-finder/src/main/java/org/datacleaner/reference/SimpleStringPattern.java

```diff
@@ -43,77 +45,95 @@
  */
 public final class SimpleStringPattern extends AbstractReferenceData implements StringPattern {
 
-	private static final long serialVersionUID = 1L;
-	private final String _expression;
-	private transient TokenPattern _tokenPattern;
-	private transient DefaultTokenizer _tokenizer;
-	private transient TokenizerConfiguration _configuration;
-	
-	public SimpleStringPattern(String name, String expression) {
-	    this(name, expression, new TokenizerConfiguration());
-	}
-
-	public SimpleStringPattern(String name, String expression, TokenizerConfiguration configuration) {
-		super(name);
-		_expression = expression;
-		_configuration = configuration;
-	}
-
-	private void readObject(ObjectInputStream stream) throws IOException, ClassNotFoundException {
-		ReadObjectBuilder.create(this, SimpleStringPattern.class).readObject(stream);
-	}
-
-	@Override
-	protected void decorateIdentity(List<Object> identifiers) {
-		super.decorateIdentity(identifiers);
-		identifiers.add(_expression);
-	}
-
-	private Tokenizer getTokenizer() {
-		if (_tokenizer == null) {
-			_tokenizer = new DefaultTokenizer(getConfiguration());
-		}
-		return _tokenizer;
-	}
-
-	private TokenizerConfiguration getConfiguration() {
-		if (_configuration == null) {
-			// TODO: Ideally we should provide all the configuration options in
-			// the constructor
-			_configuration = new TokenizerConfiguration();
-		}
-		return _configuration;
-	}
-
-	private TokenPattern getTokenPattern() {
-		if (_tokenPattern == null) {
-		    final String expression;
-		    if (LabelUtils.NULL_LABEL.equals(_expression)) {
-		        expression = null;
-		    } else if (LabelUtils.BLANK_LABEL.equals(_expression)) {
+    private static final long serialVersionUID = 1L;
+
+    private final String _expression;
+    private transient TokenPattern _tokenPattern;
+    private transient DefaultTokenizer _tokenizer;
+    private transient TokenizerConfiguration _configuration;
+
+    public SimpleStringPattern(String name, String expression) {
+        this(name, expression, new TokenizerConfiguration());
+    }
+
+    public SimpleStringPattern(String name, String expression, TokenizerConfiguration configuration) {
+        super(name);
+        _expression = expression;
+        _configuration = configuration;
+    }
+
+    private void readObject(ObjectInputStream stream) throws IOException, ClassNotFoundException {
+        ReadObjectBuilder.create(this, SimpleStringPattern.class).readObject(stream);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```

Comment:
Not implementing hashCode() when implementing equals() is breaking the contract of hashCode(); If two objects are equal, they must also return the same hashcode.